### PR TITLE
citra-qt: Improve pica command list widget

### PR DESCRIPTION
**Pull request by Lectem**
I seperated the data column into "Register" and "New Value" for the sake of clarity.
I also switched to using a QTableView because it is godamn faster, scrolling with a QTreeView with a lot of elements is extremely slow.

### CODE REVIEW

**Comment by neobrain** [comment]
QTableView is the wrong widget to use here, since we want a list-like look rather than a table-like one. Have you considered enabling http://doc.qt.io/qt-4.8/qtreeview.html#uniformRowHeights-prop instead, though?

**Comment by Lectem** [comment]
It seems that the speedup was indeed caused by the rows being the same size, unifomRowHeights did the trick for QTreeView

**Comment by neobrain** [comment]
Please improve your commit messages.

**Comment by neobrain** [comment]
Is there any actual point in displaying the raw command id? It can be looked up easily in pica.h and hence only makes for one redundant column. Instead, I suggest displaying the raw command id in the first column iff `Pica::Regs::GetCommandName(cmd.cmd_id).isEmpty()`.

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 1]
```diff
@@ -298,7 +301,13 @@ GPUCommandListWidget::GPUCommandListWidget(QWidget* parent) : QDockWidget(tr("Pi
     list_widget = new QTreeView;
     list_widget->setModel(model);
     list_widget->setFont(QFont("monospace"));
-    list_widget->setRootIsDecorated(false);
+    list_widget->setUniformRowHeights(true);
+
+#if QT_VERSION >= QT_VERSION_CHECK(5,0,0)
+    list_widget->header()->setSectionResizeMode(QHeaderView::ResizeToContents);
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: What is the difference to setResizeMode?

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 2]
```diff
@@ -298,7 +301,13 @@ GPUCommandListWidget::GPUCommandListWidget(QWidget* parent) : QDockWidget(tr("Pi
     list_widget = new QTreeView;
     list_widget->setModel(model);
     list_widget->setFont(QFont("monospace"));
-    list_widget->setRootIsDecorated(false);
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: Why was this removed?

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 3]
```diff
@@ -207,10 +207,13 @@ QVariant GPUCommandListModel::headerData(int section, Qt::Orientation orientatio
...
     {
-        if (section == 0) {
+        switch (section) {
+        case 0 :
             return tr("Command Name");
-        } else if (section == 1) {
-            return tr("Data");
+        case 1 :
+            return tr("Register");
+        case 2 :
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: redundant space before `:`

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 4]
```diff
@@ -187,14 +188,13 @@ QVariant GPUCommandListModel::data(const QModelIndex& index, int role) const {
...
-            return content;
-        } else if (index.column() == 1) {
-            QString content = QString("%1 ").arg(cmd.hex, 8, 16, QLatin1Char('0'));
-            content.append(QString("%1 ").arg(val, 8, 16, QLatin1Char('0')));
-            return content;
+        switch ( index.column() ) {
+        case 0 :
+            return QString::fromLatin1(Pica::Regs::GetCommandName(cmd.cmd_id).c_str()).append(" ");
+        case 1 :
+            return QString("%1 ").arg(cmd.cmd_id, 4, 16, QLatin1Char('0'));
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: Redundant space

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 5]
```diff
@@ -187,14 +188,13 @@ QVariant GPUCommandListModel::data(const QModelIndex& index, int role) const {
...
-            QString content = QString::fromLatin1(Pica::Regs::GetCommandName(cmd.cmd_id).c_str());
-            content.append(" ");
-            return content;
-        } else if (index.column() == 1) {
-            QString content = QString("%1 ").arg(cmd.hex, 8, 16, QLatin1Char('0'));
-            content.append(QString("%1 ").arg(val, 8, 16, QLatin1Char('0')));
-            return content;
+        switch ( index.column() ) {
+        case 0 :
+            return QString::fromLatin1(Pica::Regs::GetCommandName(cmd.cmd_id).c_str()).append(" ");
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: why append a space?

**Comment by Lectem** [comment]
Well it's faster to look for those with the ID on 3dbrew, and not all users have citra's source under their hands. Plus it assumes that all the names are correct, which might not always be the case.

File: src/citra_qt/debugger/graphics_cmdlists.cpp (hunk 5 above)
Comment (Lectem) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: Well, it was there so I didnt remove it, I guessed their was a good reason for why it was there

File: src/citra_qt/debugger/graphics_cmdlists.cpp (hunk 2 above)
Comment (Lectem) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: My bad, forgot to add it back when using QTreeView again

File: src/citra_qt/debugger/graphics_cmdlists.cpp (hunk 1 above)
Comment (Lectem) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: setResizeMode was renamed to setSectionResizeMode in Qt 5.0

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 6]
```diff
@@ -187,14 +188,13 @@ QVariant GPUCommandListModel::data(const QModelIndex& index, int role) const {
...
-            content.append(QString("%1 ").arg(val, 8, 16, QLatin1Char('0')));
-            return content;
+        switch ( index.column() ) {
+        case 0:
+            return QString::fromLatin1(Pica::Regs::GetCommandName(cmd.cmd_id).c_str());
+        case 1:
+            return QString("%1").arg(cmd.cmd_id, 4, 16, QLatin1Char('0'));
+        case 2:
+            return QString("%1").arg(val, 8, 16, QLatin1Char('0'));
         }
```
Comment (neobrain) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: Will using a `switch` here introduce warnings about missing default case?

**Comment by neobrain** [comment]
LGTM other than the potential introduction of two warnings related to unhandled `switch` cases.

**Comment by Lectem** [comment]
Visual studio 2013 didn't give me any warning (I think most compilers only do so on enums, you are not warned when using an integer in a switch)

**Comment by Subv** [comment]
```
Is there any actual point in displaying the raw command id? It can be looked up easily in pica.h and hence only makes for one redundant column. Instead, I suggest displaying the raw command id in the first column iff Pica::Regs::GetCommandName(cmd.cmd_id).empty().
```

I disagree, searching for the command id is not trivial in Pica.h, you have to look for a point of reference at the end of the file then backtrack until you find what you're looking for. Displaying the command id along with the name is better in my opinion.

**Comment by neobrain** [comment]
> I disagree, searching for the command id is not trivial in Pica.h, you have to look for a point of reference at the end of the file then backtrack until you find what you're looking for. Displaying the command id along with the name is better in my opinion.

Are you thinking of something else maybe? You literally just have to look up the name shown in the widget in the list at https://github.com/citra-emu/citra/blob/master/src/video_core/pica.h#L1005 (and then do some basic arithmetic if it e.g. shows "framebuffer+4").

**Comment by yuriks** [comment]
I would also prefer to keep the raw addresses always. The last thing I need when trying to RE something is to have to keep looking up crap.

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 7]
```diff
@@ -187,14 +188,13 @@ QVariant GPUCommandListModel::data(const QModelIndex& index, int role) const {
...
-            return content;
-        } else if (index.column() == 1) {
-            QString content = QString("%1 ").arg(cmd.hex, 8, 16, QLatin1Char('0'));
-            content.append(QString("%1 ").arg(val, 8, 16, QLatin1Char('0')));
-            return content;
+        switch ( index.column() ) {
+        case 0:
+            return QString::fromLatin1(Pica::Regs::GetCommandName(cmd.cmd_id).c_str());
+        case 1:
+            return QString("%1").arg(cmd.cmd_id, 4, 16, QLatin1Char('0'));
```
Comment (yuriks) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: Set the width to 3, since regs only go to ~300.

File: src/citra_qt/debugger/graphics_cmdlists.cpp [hunk 8]
```diff
@@ -299,6 +302,13 @@ GPUCommandListWidget::GPUCommandListWidget(QWidget* parent) : QDockWidget(tr("Pi
     list_widget->setModel(model);
     list_widget->setFont(QFont("monospace"));
     list_widget->setRootIsDecorated(false);
+    list_widget->setUniformRowHeights(true);
+
+#if QT_VERSION >= QT_VERSION_CHECK(5,0,0)
+    list_widget->header()->setSectionResizeMode(QHeaderView::ResizeToContents);
+#else
+    list_widget->header()->setResizeMode(QHeaderView::ResizeToContents);
+#endif
```
Comment (yuriks) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: "setResizeMode" is deprecated, but afaik still present in Qt 5. I'd just keep using it rather than having an ifdef.

Comment (Lectem) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: It isn't available (at least for 5.5)

Comment (yuriks) [inline src/citra_qt/debugger/graphics_cmdlists.cpp]: Hm, it's still listed in the documentation: http://doc.qt.io/qt-5/qheaderview-obsolete.html#setResizeMode

But well, it's not a big issue either way.

**Comment by yuriks** [comment]
LGTM
